File: drivers/sdk/driver.py
"""QuarkGate driver SDK helpers for Python contributors."""

from __future__ import annotations

import json
from typing import Any, Dict, Optional


def downstream(
    url: str,
    method: str,
    headers: Dict[str, str],
    body: Optional[Any],
    streaming: bool,
    provider: str,
    operation: str,
    estimate_micro: int,
) -> Dict[str, Any]:
    out: Dict[str, Any] = {
        "url": url,
        "method": method,
        "headers": headers,
        "streaming": streaming,
        "provider": provider,
        "operation": operation,
        "estimate_micro": estimate_micro,
    }
    if body is not None:
        out["body"] = body
    return out


def merge_usage(base: Dict[str, Any], patch: Dict[str, Any]) -> Dict[str, Any]:
    out = dict(base or {})
    out.update(patch or {})
    return out

File: drivers/sdk/driver.js
/**
 * QuarkGate Driver SDK types and helpers.
 */

function envelope(provider, operation, payload, hints) {
  return {
    quarkgate_version: '1',
    provider,
    operation,
    payload,
    metering_hints: hints,
  };
}

function downstream(url, method, headers, body, streaming, provider, operation, estimateMicro) {
  return {
    url,
    method,
    headers,
    body,
    streaming,
    provider,
    operation,
    estimate_micro: estimateMicro || 0,
  };
}

/** Helpers for driver stream metering fallbacks. */
function estimateOutputTokensFromChunk(data) {
  if (!data || data === '[DONE]') {
    return 0;
  }
  return Math.ceil(String(data).length / 4);
}

module.exports = { envelope, downstream, estimateOutputTokensFromChunk };
